### PR TITLE
DAOS-9636 object: catch all retry error for update and punch ops

### PR DESCRIPTION
**Pull request by wangdi1**
If retry error is caught before obj_comp_cb() callback is
registered, even though obj_comp_cb() is executed, but the
following tse_task_complete() will cancel the retry callback.

So cleanup the tx convert process to catch all retry failure
correctly.

Add test case to verify it.

Signed-off-by: Di Wang <di.wang@intel.com>

### CODE REVIEW

File: src/object/cli_obj.c [hunk 1]
```diff
@@ -4386,6 +4389,9 @@ obj_comp_cb(tse_task_t *task, void *data)
 		     task->dt_result == -DER_CSUM)))
 			obj_auxi->io_retry = 0;
 
+		if (task->dt_result == -DER_SHARDS_OVERLAP)
```
Comment (Nasf-Fan) [inline src/object/cli_obj.c:4392]: Under which cases the DER_SHARDS_OVERLAP will be returned via obj_comp_cb()? As my understand, during obj_req_get_tgts(), we will check whether there are two shares locate on the same target or not. If we pass such check, then where the -DER_SHARDS_OVERLAP will returned later?

Comment (wangdi1) [reply src/object/cli_obj.c:4392]: yes, obj_req_get_tgts() will return OVERLAP if there are two shards on the same target, then obj_comp_cb() will be called and tx_convert is bet set there, which will trigger retry, then go into dc_tx_convert.

In original implementation, it did sth like

if (error) {
             obj_comp_cb();
              tse_task_complete();
}

Even obj_comp_cb() may see the task needs to retry, and re-init the task, but tse_task_complete() may still terminate the task.

**Review by Nasf-Fan — APPROVED** [review]

File: src/object/cli_obj.c [hunk 2]
```diff
@@ -5024,66 +5035,40 @@ dc_obj_update(tse_task_t *task, struct dtx_epoch *epoch, uint32_t map_ver,
...
-			}
-			goto out_task;
-		}
+	if (rc != 0)
+		D_GOTO(out_task, rc);
 
-		if (obj_auxi->is_ec_obj && obj_auxi->req_reasbed) {
-			args->iods = obj_auxi->reasb_req.orr_uiods;
-			args->sgls = obj_auxi->reasb_req.orr_usgls;
-		}
```
Comment (liuxuezhao) [inline src/object/cli_obj.c:5070]: the above 4 lines was for the case that obj_rw_req_reassemb() executed above so it replaced args->iods/sgls as reassembled iods/sgls, but the dc_tx_convert() will reassemble then again, so here assign back the original user iods/sgls.
now it is removed, but I did not see where with such similar handling? 


In the new code, for example an new obj update task, it call obj_rw_req_reassemb() and then call obj_req_get_tgts() which returns -DER_SHARDS_OVERLAP, then its complete cb (obj_comp_cb) called that will set tx_convert flag, two things looks not clear to me -
1) how can that task be retried? in obj_comp_cb() it just set the tx_convert flag, but seems will not be retried?
2) when that task get retried, above L5045 will call dc_tx_convert() but the "args->iods/sgls" was set as reassembled iod/sgls, how can it reassemble it again in dc_tx_convert()?

**Review by liuxuezhao — APPROVED** [review]

File: src/object/cli_obj.c (hunk 2 above)
Comment (wangdi1) [reply src/object/cli_obj.c:5070]: you are right, this should be set back before convert.

**Review by Nasf-Fan — APPROVED** [review]

**Pull request merged**